### PR TITLE
[SMTChecker] Support StringLiteral type

### LINKED ISSUE #7165: [SMTChecker] StringLiteral causes type mismatch
From https://github.com/ethereum/solidity/issues/7011

```
pragma experimental SMTChecker;
  
contract C {
    function f(bytes32 _x) public pure {
        require(_x != "");
    }
}
```

StringLiterals are currently supported as Arrays, following `bytes`. Since StringLiterals can also be assigned/compared to `bytesNN` which is supported as `Int`, this causes a type mismatch inside the solver.
The fix for this involves a layer of type conversion for assignments similar to how it's done in codegen.

### PR DESCRIPTION
**Pull request by leonardoalt**
Fixes #7165 

Depends on #7208

### CODE REVIEW

File: libsolidity/formal/SymbolicVariables.cpp [hunk 1]
```diff
@@ -171,27 +180,45 @@ SymbolicMappingVariable::SymbolicMappingVariable(
...
+Expression SymbolicArrayVariable::currentValueAsType(solidity::TypePointer const& _targetType) const
+{
+	solAssert(_targetType, "");
+	// StringLiterals are encoded as SMT arrays in the generic case,
+	// but they can also be compared/assigned to fixed bytes, in which
+	// case they'd need to be encoded as numbers.
+	if (m_originalType->category() == solidity::Type::Category::StringLiteral)
+		if (_targetType->category() == solidity::Type::Category::FixedBytes)
+		{
+			auto strType = dynamic_cast<StringLiteralType const*>(m_originalType);
```
Comment (ekpyron) [inline libsolidity/formal/SymbolicVariables.cpp]: You could just have used this in the outer ``if`` (the ``dynamic_cast`` will return ``nullptr``, if the category is wrong).

File: libsolidity/formal/SymbolicVariables.h [hunk 2]
```diff
@@ -52,6 +53,7 @@ class SymbolicVariable
 	virtual ~SymbolicVariable() = default;
 
 	Expression currentValue() const;
+	virtual Expression currentValueAsType(solidity::TypePointer const& _targetType) const;
```
Comment (ekpyron) [inline libsolidity/formal/SymbolicVariables.h]: Not sure about the name of this... sounds a bit like you return a type that encodes the current value...
What about making ``currentValue`` virtual instead and giving it a ``solidity::TypePointer const& _targetType = TypePointer{}`` as argument? You can still return ``SymbolicVariable::currentValueAsType({});`` in the override in ``SymbolicArrayValue`` where you now return ``currentValue``...

Or alternatively there might be a better name for this...

File: libsolidity/formal/SymbolicVariables.cpp [hunk 3]
```diff
@@ -95,17 +102,18 @@ SymbolicBoolVariable::SymbolicBoolVariable(
 	string _uniqueName,
 	EncodingContext& _context
 ):
-	SymbolicVariable(move(_type), move(_uniqueName), _context)
+	SymbolicVariable(_type, move(_type), move(_uniqueName), _context)
```
Comment (ekpyron) [inline libsolidity/formal/SymbolicVariables.cpp]: Remove move.

File: libsolidity/formal/SymbolicVariables.cpp [hunk 4]
```diff
@@ -95,17 +102,18 @@ SymbolicBoolVariable::SymbolicBoolVariable(
...
 }
 
 SymbolicIntVariable::SymbolicIntVariable(
 	solidity::TypePointer _type,
+	solidity::TypePointer _originalType,
 	string _uniqueName,
 	EncodingContext& _context
 ):
-	SymbolicVariable(move(_type), move(_uniqueName), _context)
+	SymbolicVariable(move(_type), move(_originalType), move(_uniqueName), _context)
```
Comment (ekpyron) [inline libsolidity/formal/SymbolicVariables.cpp]: Could remove all the moves in these cases.

**Review by ekpyron — COMMENTED** [review]
Looks good, but I'd still refactor the ``currentValueAsType`` bit a bit.

**Comment by leonardoalt** [comment]
@ekpyron updated

**Comment by leonardoalt** [comment]
@ekpyron updated `SMTEncoder::expr(...)`

File: libsolidity/formal/SymbolicVariables.cpp [hunk 5]
```diff
@@ -171,27 +175,41 @@ SymbolicMappingVariable::SymbolicMappingVariable(
...
 
+Expression SymbolicArrayVariable::currentValue(solidity::TypePointer const& _targetType) const
+{
+	if (_targetType)
+		// StringLiterals are encoded as SMT arrays in the generic case,
+		// but they can also be compared/assigned to fixed bytes, in which
+		// case they'd need to be encoded as numbers.
+		if (auto strType = dynamic_cast<StringLiteralType const*>(m_originalType))
+			if (_targetType->category() == solidity::Type::Category::FixedBytes)
+				return smt::Expression(u256(toHex(asBytes(strType->value()), HexPrefix::Add)));
```
Comment (leonardoalt) [inline libsolidity/formal/SymbolicVariables.cpp:202]: Any shortcut to this? @chriseth

Comment (ekpyron) [reply libsolidity/formal/SymbolicVariables.cpp:202]: Maybe ``import_bits`` works? https://www.boost.org/doc/libs/1_66_0/libs/multiprecision/doc/html/boost_multiprecision/tut/import_export.html

Comment (leonardoalt) [reply libsolidity/formal/SymbolicVariables.cpp:202]: hmm not sure, still need an `u256` in the end

Comment (ekpyron) [reply libsolidity/formal/SymbolicVariables.cpp:202]: An ``u256`` is a boost multiprecision number - it should work as first argument to ``import_bits``...

Comment (leonardoalt) [reply libsolidity/formal/SymbolicVariables.cpp:202]: After talking to @ekpyron and @chriseth I'm convinced about the way it is at the moment. @ekpyron objections?

Comment (ekpyron) [reply libsolidity/formal/SymbolicVariables.cpp:202]: Yeah - it feels a bit wasteful to first convert binary to hex only to have ``u256`` convert it back to binary again, but it's fine.

**Review by ekpyron — APPROVED** [review]

**Pull request merged**